### PR TITLE
Add CronJob and StatefulSet targetting when copy_target is enabled.

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: [#404](https://github.com/metalbear-co/operator/issues/404) and [#543](https://github.com/metalbear-co/operator/issues/543)

1. Adds `CronJob` as a possible `target` (requires `copy_target` feature to be enabled);
2. Adds `StatefulSet` as a possible `target` (requires `copy_target` feature to be enabled);

### CODE REVIEW

File: mirrord/config/src/target/cron_job.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            f,
+            "{}{}",
+            self.container
+                .as_ref()
+                .map(|c| format!("{c}/"))
+                .unwrap_or_default(),
+            self.cron_job.clone()
+        )
+    }
+}
```
Comment (Razz4780) [inline mirrord/config/src/target/cron_job.rs:29]: I don't think this impl is needed, most other target types (excluding `JobTarget`) don't have it :eyes: 
Could you either remove it or change the order of elements (right now the container is displayed before the cron job name). Applies to both `CronJobTarget` and `JobTarget`

**Review by Razz4780 — CHANGES_REQUESTED** [review]
There's also one thing I missed when reviewing previous `JobTarget` PR - we need a feature flag in the operator status to check whether this target type is supported

**Comment by Razz4780** [comment]
> * Issue: [#404](https://github.com/metalbear-co/operator/issues/404)

Issue not found ;/

**Comment by aviramha** [comment]
> There's also one thing I missed when reviewing previous `JobTarget` PR - we need a feature flag in the operator status to check whether this target type is supported

We could trust the `mirrord ls` in terms of UX, no?

**Comment by Razz4780** [comment]
> > There's also one thing I missed when reviewing previous `JobTarget` PR - we need a feature flag in the operator status to check whether this target type is supported
> 
> We could trust the `mirrord ls` in terms of UX, no?

Yes, job targets should not be listed unless operator supports them

**Comment by aviramha** [comment]
> > > There's also one thing I missed when reviewing previous `JobTarget` PR - we need a feature flag in the operator status to check whether this target type is supported
> > 
> > 
> > We could trust the `mirrord ls` in terms of UX, no?
> 
> Yes, job targets should not be listed unless operator supports them

I'm saying we don't need a feature flag because it either returns it as an option or it doesn't.

**Comment by aviramha** [comment]
If we want a feature flag for nicer error, I'd make a "target not supported" error with appropriate message and re-use it, as we're adding more and more targets, and it's a bit weird to have a feature flag for each

**Comment by Razz4780** [comment]
> If we want a feature flag for nicer error, I'd make a "target not supported" error with appropriate message and re-use it, as we're adding more and more targets, and it's a bit weird to have a feature flag for each

Okok, makes sense

File: mirrord/config/src/target.rs [hunk 2]
```diff
@@ -402,6 +419,12 @@ impl CollectAnalytics for &TargetConfig {
                         flags |= TargetAnalyticFlags::CONTAINER;
                     }
                 }
+                Target::CronJob(cron_job) => {
+                    flags |= TargetAnalyticFlags::JOB;
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/config/src/target.rs:423]: Changed.

File: mirrord/config/src/target/cron_job.rs (hunk 1 above)
Comment (meowjesty) [reply mirrord/config/src/target/cron_job.rs:29]: Removed from both places.

File: mirrord/config/src/lib.rs [hunk 3]
```diff
@@ -389,7 +389,7 @@ impl LayerConfig {
                 .target
                 .path
                 .as_ref()
-                .map(|target| matches!(target, target::Target::Job(_)))
+                .map(|target| matches!(target, target::Target::Job(_) | target::Target::CronJob(_)))
```
Comment (Razz4780) [inline mirrord/config/src/lib.rs:392]: `StatefulSet` as well

File: mirrord/config/src/target.rs [hunk 4]
```diff
@@ -184,6 +188,7 @@ mirrord-layer failed to parse the provided target!
     >> deploy/<deployment-name>[/container/container-name]
     >> pod/<pod-name>[/container/container-name]
     >> job/<job-name>[/container/container-name]
+    >> cronjob/<cronjob-name>[/container/container-name]
```
Comment (Razz4780) [inline mirrord/config/src/target.rs:191]: `StatefulSet` as well

File: mirrord/config/src/target.rs [hunk 5]
```diff
@@ -207,6 +212,7 @@ mirrord-layer failed to parse the provided target!
 /// - `container/{sample-container}`;
 /// - `containername/{sample-container}`.
 /// - `job/{sample-job}`;
+/// - `cronjob/{sample-job}`;
```
Comment (Razz4780) [inline mirrord/config/src/target.rs:215]: `StatefulSet` as well

File: mirrord/kube/src/api/runtime/stateful_set.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    fn container(&self) -> Option<&str> {
+        self.container.as_deref()
+    }
+
+    async fn get_labels(resource: &Self::Resource) -> Result<BTreeMap<String, String>> {
+        resource
+            .spec
+            .as_ref()
+            .and_then(|spec| spec.selector.match_labels.clone())
+            .ok_or_else(|| KubeApiError::missing_field(resource, ".spec.selector.match_labels"))
```
Comment (Razz4780) [inline mirrord/kube/src/api/runtime/stateful_set.rs:25]:
```suggestion
            .ok_or_else(|| KubeApiError::missing_field(resource, ".spec.selector.matchLabels"))
```
It's the original name used in yaml

**Review by Razz4780 — CHANGES_REQUESTED** [review]

**Review by Razz4780 — APPROVED** [review]

**Pull request merged**